### PR TITLE
Issue #859 Define Server Side Cache Configuration Contents and Compat…

### PR DESCRIPTION
**Pull request by gibsong**
…ibility Rules

### CODE REVIEW

**Comment by ljacomet** [comment]
This PR lacks testing from the Ehcache API level which would show that our exception handling in the face of `auto-create` is wrong.

The way `org.ehcache.clustered.client.internal.service.DefaultClusteringService#getServerStoreProxy` works is a problem as validation failure will always be hidden by creation failures in `auto-create` mode.

**Comment by cljohnso** [comment]
@ljacomet, I believe your assertion is that a validation failure with `auto-create` is hidden by a creation failure isn't correct.  The exception that gets returned from the server in this case is a `ClusteredStoreValidationException` (which subclasses `RuntimeException`).  That `RuntimeException` is rethrown from `EhcacheClientEntity.validateServerStore` and is emitted from `DefaultServerStore.getServerStoreProxy`.  (This is confirmed by a couple of tests in `DefaultServerStoreProxyTest`.)  We probably need to have it wrapped in `EhcacheClientEntity` so we get the full client-side stack trace in the exception -- I think the way it is now only gets the server-side stack trace -- but I believe the test path you seek is covered.

**Comment by cljohnso** [comment]
I looked at the tests in `DefaultClusteringServiceTest` again and they don't actually exercise the logic in `ServerStoreCompatibility` ... some tests need to be added in `DefaultClusteringServiceTest` (or a supplementary class) to ensure proper operation/visibility of validation failures.

**Comment by cljohnso** [comment]
I cobbled together a sample and @ljacomet is right ... the `ClusteredStoreValidationException` is hidden under a `CachePersistenceException` by `EhcacheClientEntity.invokeInternal`.  A scattering of changes in `DefaultClusteringService` and `EhcacheClientEntity` are needed to properly expose the `ClusteredStoreValidationException`.  We should discuss ...

**Comment by cljohnso** [comment]
@ljacomet indicated issue #1130 should be handled in this PR -- this necessitates some changes in `ServerStoreCompatibility` to ensure the server-side and client-specified pool allocations are of the same type before casting.

**Comment by gibsong** [comment]
I was able to expose the ClusteredStoreValidationException via a unit test.  I still need to write some more tests to see if there are any other bugs.  A discussion would help find additional cases I might not be seeing.

**Comment by gibsong** [comment]
ok I can combine #1130 with this PR.  I have already made the fix to prevent a user from changing a PoolAllocation from fixed to shared and vice versa.

File: clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreCompatibility.java [hunk 1]
```diff
@@ -70,12 +73,24 @@ public void verify(ServerStoreConfiguration serverConfiguration, ServerStoreConf
...
     isCompatible &= compareField(sb, "valueSerializerType", serverConfiguration.getValueSerializerType(), clientConfiguration.getValueSerializerType());
+    isCompatible &= compareConsistencyField(sb, serverConfiguration.getConsistency(), clientConfiguration.getConsistency());
 
     if (!isCompatible) {
       throw new ClusteredStoreValidationException(sb.toString());
     }
   }
 
+  private static boolean compareConsistencyField(StringBuilder sb, Consistency serverConsistencyValue, Consistency clientConsistencyValue)
+  {
```
Comment (ljacomet) [inline clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreCompatibility.java]: Code formatting mismatch

**Comment by ljacomet** [comment]
Please squash some of these commits - goal is to have merged commits that enable `git bisect`, that is at each commit, tests should pass.

File: clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreCompatibility.java [hunk 2]
```diff
@@ -70,12 +73,24 @@ public void verify(ServerStoreConfiguration serverConfiguration, ServerStoreConf
     isCompatible &= compareField(sb, "actualValueType", serverConfiguration.getActualValueType(), clientConfiguration.getActualValueType());
     isCompatible &= compareField(sb, "keySerializerType", serverConfiguration.getKeySerializerType(), clientConfiguration.getKeySerializerType());
     isCompatible &= compareField(sb, "valueSerializerType", serverConfiguration.getValueSerializerType(), clientConfiguration.getValueSerializerType());
+    isCompatible &= compareConsistencyField(sb, serverConfiguration.getConsistency(), clientConfiguration.getConsistency());
```
Comment (lorban) [inline clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreCompatibility.java]: There doesn't seem to be a test for this check

File: clustered/common/src/test/java/org/ehcache/clustered/common/ServerStoreCompatibilityTest.java [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+import org.ehcache.clustered.common.ServerStoreConfiguration.PoolAllocation.Fixed;
+import org.ehcache.clustered.common.ServerStoreConfiguration.PoolAllocation.Shared;
+import static org.junit.Assert.fail;
+import org.junit.Test;
+
+/**
+ *
+ * @author GGIB
+ */
+public class ServerStoreCompatibilityTest
```
Comment (lorban) [inline clustered/common/src/test/java/org/ehcache/clustered/common/ServerStoreCompatibilityTest.java]: You should observe the existing code formatting. There is some freedom but not around curly braces placement.

File: clustered/common/src/test/java/org/ehcache/clustered/common/ServerStoreCompatibilityTest.java [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+    ServerStoreCompatibility serverStoreCompatibility = new ServerStoreCompatibility();
+
+    try
+    {
+      serverStoreCompatibility.verify(serverConfiguration, clientConfiguration);
+      fail("Expected ClusteredStoreValidationException");
+    }
+    catch(ClusteredStoreValidationException e)
+    {
+      System.out.println("errorMsg: " + e.getMessage());
```
Comment (lorban) [inline clustered/common/src/test/java/org/ehcache/clustered/common/ServerStoreCompatibilityTest.java]: No `System.out.println()` please.

File: clustered/common/src/test/java/org/ehcache/clustered/common/ServerStoreCompatibilityTest.java [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+      Assert.assertTrue("test failed", e.getMessage().equals(ERROR_MESSAGE_BASE + "storedKeyType existing: " + serverConfiguration.getStoredKeyType() + ", desired: " + clientConfiguration.getStoredKeyType()));
```
Comment (lorban) [inline clustered/common/src/test/java/org/ehcache/clustered/common/ServerStoreCompatibilityTest.java]: Better use `assertThat(... , is(true))`

File: clustered/common/src/test/java/org/ehcache/clustered/common/ServerStoreCompatibilityTest.java [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+
+    ServerStoreCompatibility serverStoreCompatibility = new ServerStoreCompatibility();
+
+    try
+    {
+      serverStoreCompatibility.verify(serverConfiguration, clientConfiguration);
+    }
+    catch(Exception e)
+    {
+      fail("No Exception Expected!");
```
Comment (lorban) [inline clustered/common/src/test/java/org/ehcache/clustered/common/ServerStoreCompatibilityTest.java]: It's much better to declare the test method with `throws Exception` than catching it and failing the test without knowing what the exception was.

File: clustered/client/src/test/java/org/ehcache/clustered/client/internal/service/DefaultClusteringServiceTest.java [hunk 7]
```diff
@@ -1350,6 +1352,463 @@ public void testFullDestroyAll() throws Exception {
...
+        .resourcePool("sharedPrimary", 16, MemoryUnit.MB, "serverResource1")
+        .resourcePool("sharedSecondary", 16, MemoryUnit.MB, "serverResource2")
+        .resourcePool("sharedTertiary", 32, MemoryUnit.MB)
+        .build();
+    DefaultClusteringService creationServiceGood = new DefaultClusteringService(creationConfigGood);
+
+    try {
+      creationServiceGood.start(null);
+    } catch (Exception e) {
+      fail("There should not be an error on start");
```
Comment (lorban) [inline clustered/client/src/test/java/org/ehcache/clustered/client/internal/service/DefaultClusteringServiceTest.java]: Make the method `throws Exception` instead of catching and failing.

**Comment by gibsong** [comment]
I checked in this code but had some failing tests.  They appeared to be only failing on Windows and were not related to my code changes.  They were failing on the master branch.

The following tests were failing:
1.
class: EhcacheSettingsProviderTestTest
test: test_standalone_ehcache()

2. (i had to comment out the above test and then this failed)
class: TerminatedServerTest
test: testTerminationBeforeCacheManagerClose()

File: clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreCompatibility.java (hunk 2 above)
Comment (gibsong) [inline clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreCompatibility.java]: I actually have a test in ServerStoreCompatibilityTest, it's called testConsitencyMismatch()

File: clustered/common/src/test/java/org/ehcache/clustered/common/ServerStoreCompatibilityTest.java (hunk 3 above)
Comment (gibsong) [inline clustered/common/src/test/java/org/ehcache/clustered/common/ServerStoreCompatibilityTest.java]: sorry, bad habit formatting that way.  I changed it

File: clustered/common/src/test/java/org/ehcache/clustered/common/ServerStoreCompatibilityTest.java (hunk 4 above)
Comment (gibsong) [inline clustered/common/src/test/java/org/ehcache/clustered/common/ServerStoreCompatibilityTest.java]: its been removed

File: clustered/common/src/test/java/org/ehcache/clustered/common/ServerStoreCompatibilityTest.java (hunk 5 above)
Comment (gibsong) [inline clustered/common/src/test/java/org/ehcache/clustered/common/ServerStoreCompatibilityTest.java]: I changed to check to use assertThat().  Why is this better?  Sorry this is my first time with junit.  More familiar with testng.

File: clustered/common/src/test/java/org/ehcache/clustered/common/ServerStoreCompatibilityTest.java (hunk 6 above)
Comment (gibsong) [inline clustered/common/src/test/java/org/ehcache/clustered/common/ServerStoreCompatibilityTest.java]: you are right.  I made the change.

File: clustered/client/src/test/java/org/ehcache/clustered/client/internal/service/DefaultClusteringServiceTest.java (hunk 7 above)
Comment (gibsong) [inline clustered/client/src/test/java/org/ehcache/clustered/client/internal/service/DefaultClusteringServiceTest.java]: i updated it as requested.  Nice catch.

File: .gitattributes [hunk 8]
```diff
@@ -1,6 +1,8 @@
 # Set the default behavior to LF, because checkstyle enforces it
 * text eol=lf
 
+*.bat eol=crlf
```
Comment (ljacomet) [inline .gitattributes]: No idea why these changes appear here but they should not. They are already on master following the Gradle 2.14 upgrade

Comment (gibsong) [inline .gitattributes]: This pull request was accidentally closed. I then pushed my PR changes, which included the Gradle upgrade, and re-opened the PR.  Could this be why they are showing?  Maybe it did a diff from the closed PR code to the new re-opened PR code.

File: clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreCompatibility.java (hunk 1 above)
Comment (gibsong) [inline clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreCompatibility.java]: fixed formatting
